### PR TITLE
Fixing thrown exception when 'noRetry' flag is set. #8457

### PR DESCRIPTION
**Pull request by ahmedhagii**
**Behaviour:**
As discussed in the issue #8457, an exception was being thrown whenever there's a queue of methods with 'noRetry' flag set and connection is dropped then re-established.

**Cause:**
It was due to a wrong way of deleting elements from an array while iterating over it leading to the access of out of bounds elements.

**Steps:**
1) I made the old test aware of the problem by sending 2 methods instead of 1 so the loop will continue on after deleting the first element.
2) The test was failing with the old code.
3) Instead of looping and using "splice" to delete elements, I used builtin "filter" to filter out methods not satisfying the condition while notifying those methods their invocation has failed.
4) Tests now pass.

**Notes:**
Actually I didn't fully grasp how the 'StubStream' is working or how are methods added to "Stream.sent" so if someone can quickly explain it or point me somewhere I can read, it will be really helpful.

This is my first PR to an open source project, so excuse me if I missed any trivial steps 😅 or if this is not in the requested format and would appreciate any feedback.

### CODE REVIEW

File: packages/ddp-client/livedata_connection_tests.js [hunk 1]
```diff
@@ -819,34 +819,48 @@ if (Meteor.isClient) {
 
     startAndConnect(test, stream);
 
-    var methodCallbackFired = false;
-    var methodCallbackErrored = false;
+    var firstMethodCallbackFired = false;
+    var firsMethodCallbackErrored = false;
```
Comment (hwillson) [inline packages/ddp-client/livedata_connection_tests.js]: Looks like a typo here - `firsMethodCallbackErrored` should be `firstMethodCallbackErrored`.

**Comment by ahmedhagii** [comment]
Thanks @hwillson for reviewing this. I have removed that file and fixed the typo.

**Comment by benjamn** [comment]
Your tests all passed locally for me, so I manually squashed this and merged it to `devel` just now: https://github.com/meteor/meteor/commit/b59513540efe8d2f178eb32527ec31d8d06bf727. Thanks for diagnosing and fixing this issue!

**Comment by ahmedhagii** [comment]
@benjamn Thanks Ben :)

**Pull request closed without merging**